### ISSUE TITLE
func compiler trashes the stack

### ISSUE THREAD

**Issue by dosvidos** [ton-blockchain/ton#106]
The following FunC code
```
  if (cs.slice_refs()) {
    int i = -1;
    do
    {
      i = i + 1;
    } until(i > 10);
  }
  set_data(begin_cell()
    .store_uint(stored_seqno + 1, 32)
    .store_uint(N, 32)
    .store_uint(K, 32)
    .store_dict(pubkeys)
    .store_dict(orders)
    .end_cell());
```
Is incorrectly compiled into the following FIFT assembly
```
SREFS  //  orders pubkeys K stored_seqno N _79
  IF:<{  //  orders pubkeys K stored_seqno N
    -1 PUSHINT  //  orders pubkeys K stored_seqno N i=-1
    UNTILEND
    INC  //  orders pubkeys K stored_seqno N i
    DUP  //  orders pubkeys K stored_seqno N i i
    10 GTINT  //  orders pubkeys K stored_seqno N i _85
    2DROP  //  orders pubkeys K stored_seqno N
  }>  //  orders pubkeys K stored_seqno N
  SWAP  //  orders pubkeys K N stored_seqno
  INC  //  orders pubkeys K N _87
  NEWC  //  orders pubkeys K N _87 _88
  32 STU  //  orders pubkeys K N _90
  32 STU  //  orders pubkeys K _92
  32 STU  //  orders pubkeys _94
  STDICT  //  orders _95
  STDICT  //  _96
  ENDC  //  _97
```
Leading to TVM runtime error:
```
execute implicit RET
until loop body end
until loop terminated
execute SWAP
execute INC
execute NEWC
execute STU 32
execute STU 32
execute STU 32
handling exception code 7: not an integer
default exception handler, terminating vm with exit code 7
```

The full sample sources are attached to the bug in a ZIP archive.
To reproduce the error, extract it and run `fift wallet-debug.fif`
[error.zip](https://github.com/ton-blockchain/ton/files/3703363/error.zip)

**Comment by EmelyanenkoK** [comment]
Can confirm this behaviour.
Looks like problem is in `do {} until ();` construction.
This program falls into infinite loop instead of finishing after first step:
```
(cell) recv_external(slice in_msg) impure {
   cell mydict = new_dict();
   builder r = begin_cell().store_int(10, 16);
   mydict~udict_set_builder(32, 1, r );
   int index = -1;
   do {
      (var index, var value, var found) = mydict.udict_get_next?(32, index);
    } until ( ~ found );
  return mydict;
}

() recv_internal() impure {
  ;; Nothing
}
```
Full example is available in 
[error_sample.zip](https://github.com/ton-blockchain/ton/files/3708268/error_sample.zip)

Analysis shows that it this example some stack elements were not swapped. In other cases, like in original post, too many elements are popped from the stack. (Note that if you remove useless `do {} until ();` from original post code, issue goes).

**Comment by ton-blockchain** [comment]
Should be fixed in commit 38c3e39066a9c3b8a62b2f4cad54df4dc6307360. Please check.

**Comment by dosvidos** [comment]
This is indeed fixed by the designated commit. Thx.
